### ISSUE TITLE
Single frequency line problem

### ISSUE THREAD

**Issue by mayfool** [rishikksh20/iSTFTNet-pytorch#3]
Thanks for the implemention of ISTFT. It has better inference speed than hifigan v1.However, I found that there is a single frequency line which would cause little noise.I use 16KHZ dataset for training.And all the line is extractly at 4k which is the middle of the all frequency.I'm trying to fix this problem, do you have the same problem?

**Comment by rishikksh20** [comment]
yes I also saw that line but it won't impact my quality

**Comment by leminhnguyen** [comment]
Hi @mayfool, can you show the mel-spectrogram image which has a single frequency line ?

**Comment by mayfool** [comment]
> Hi @mayfool, can you show the mel-spectrogram image which has a single frequency line ?

All synthesised  wavs have single frequency line. Not few of them. So I think it has no matter with input mels.

**Comment by xiaoyangnihao** [comment]
![Uploading image.png…]()
I alse see this frequency line at 5-6khz with 18w steps , bs=16 now

**Comment by xiaoyangnihao** [comment]
> > Hi @mayfool, can you show the mel-spectrogram image which has a single frequency line ?
> 
> All synthesised wavs have single frequency line. Not few of them. So I think it has no matter with input mels.

@mayfool have you solved this problem now?

**Comment by mayfool** [comment]
> > > Hi @mayfool, can you show the mel-spectrogram image which has a single frequency line ?
> > 
> > 
> > All synthesised wavs have single frequency line. Not few of them. So I think it has no matter with input mels.
> 
> @mayfool have you solved this problem now?

@xiaoyangnihao Nope..

**Comment by Dekakhrone** [comment]
I implemented the code for any number of upsampling, and it seems that the problem with the single line frequency (which, by the way, lies in the region of 5500 Hz, that is, half of the fmax at a sample rate of 22050 Hz) is observed specifically for the configuration C8C8I. 

Below is a comparison of C8C8I and C8C8C2I

![Screenshot from 2022-07-12 12-13-41](https://user-images.githubusercontent.com/51401973/178455366-c32c0662-aa03-4784-8aa9-ffb5fd0e8863.png)
![Screenshot from 2022-07-12 12-13-16](https://user-images.githubusercontent.com/51401973/178455374-f3fd829f-3783-4390-b58d-c9dc59d95067.png)

**Comment by jstzwj** [comment]
I also encountered this problem. I set the sample rate to 48k, the horizontal line appears at 12k HZ (half of the fmax 24k HZ).

**Comment by leng-yue** [comment]
I encountered this problem too.

**Comment by ease-zh** [comment]
I have trained a model with 100k steps, it sounds good, but looking into the generated spec, it seems to make no sense.  
Looking at the generated spec, we can find the highlight in 2k and 6k Hz, which do not exist in the input mel, even not in the mel re-calc from the generated audio. Comparing the two specs, we can say, that even it can be converted to audio by istft, the model output is actually NOT spectrum.   
Moreover, run ifft and de-window for the generated spec, it generates the audio frames with length 16. For sil, the beginning 100 frames look very similar, thus, followed with the overlap-and-add method, just the same signal added with 4 samples moving behind. And that's it, the 4k Hz signal.

I think adding spec loss directly to the generated spec may be a good idea to improve the quality and to solve this question.
I will train a new model to see what will happen.
![image](https://github.com/rishikksh20/iSTFTNet-pytorch/assets/28568749/e6d4bda5-8e41-4359-a6fc-691387f5bbfb)

**Comment by ease-zh** [comment]
Yet I have trained a model with linear spec loss for 100k steps, the 4k Hz signal is gone.
Since I randomly selected the sample to syn, I cannot find the exact sample that was used for the aforementioned plot. So I re-plot the figure with a new sample.
Looking at the figure with spec loss, the generated spec looks to make sense, the 4k line was gone, and the waveform for the beginning silence is not always the same. Now everything looks all right.
![image](https://github.com/rishikksh20/iSTFTNet-pytorch/assets/28568749/c5a788d3-7e17-47b6-bc80-54322e8b59d4)
![image](https://github.com/rishikksh20/iSTFTNet-pytorch/assets/28568749/1a35bded-3f4c-4a40-95bc-9efa1ddd41d0)


> I have trained a model with 100k steps, it sounds good, but looking into the generated spec, it seems to make no sense. Looking at the generated spec, we can find the highlight in 2k and 6k Hz, which do not exist in the input mel, even not in the mel re-calc from the generated audio. Comparing the two specs, we can say, that even it can be converted to audio by istft, the model output is actually NOT spectrum. Moreover, run ifft and de-window for the generated spec, it generates the audio frames with length 16. For sil, the beginning 100 frames look very similar, thus, followed with the overlap-and-add method, just the same signal added with 4 samples moving behind. And that's it, the 4k Hz signal.
> 
> I think adding spec loss directly to the generated spec may be a good idea to improve the quality and to solve this question. I will train a new model to see what will happen. ![image](https://private-user-images.githubusercontent.com/28568749/292737138-e6d4bda5-8e41-4359-a6fc-691387f5bbfb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQxNjI5MjIsIm5iZiI6MTcwNDE2MjYyMiwicGF0aCI6Ii8yODU2ODc0OS8yOTI3MzcxMzgtZTZkNGJkYTUtOGU0MS00MzU5LWE2ZmMtNjkxMzg3ZjViYmZiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTAyVDAyMzAyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlhNGJkNzU0YzU0YzZkNjcwY2NmY2UwMDUxZDA0YzM1ODgzMTY3M2FmYmZhNjFkYzFiM2IxYmE3YTdlOGZhMDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1GFRzpeadFBsQfugjvM4rqGlHUwZPgTMkKGwuCugNpI)

**Comment by rishikksh20** [comment]
@ease-zh what changes you have made to this repo to achieve that ?

**Comment by ease-zh** [comment]
> @ease-zh what changes you have made to this repo to achieve that ?

Just a l1 loss for the generated linear spec, `loss_spec = F.l1_loss(y_spec, spec) * 45`. However, after careful comparison, I found that the spec loss harms the audio quality.   
Maybe changing the loss formula or tuning the loss weight can further improve the quality.  Do you have more suggestions?

By the way, I think using `reflection_pad` before the `conv_post` seems to make less sense, although it did work. I guess here is to justify the length so that `torch.istft` will give the exact samples as used to calculate the mel? But the difference is caused by that `torch.istft` only supports `center` mode, while calculating the mel, we manually pad the wav, and set `center=False` for `torch.stft`. This may not affect the final synthesis, and I will give it a try later.

**Comment by a897456** [comment]
> > @ease-zh what changes you have made to this repo to achieve that ?
> 
> Just a l1 loss for the generated linear spec, `loss_spec = F.l1_loss(y_spec, spec) * 45`. However, after careful comparison, I found that the spec loss harms the audio quality. Maybe changing the loss formula or tuning the loss weight can further improve the quality. Do you have more suggestions?
> 
> By the way, I think using `reflection_pad` before the `conv_post` seems to make less sense, although it did work. I guess here is to justify the length so that `torch.istft` will give the exact samples as used to calculate the mel? But the difference is caused by that `torch.istft` only supports `center` mode, while calculating the mel, we manually pad the wav, and set `center=False` for `torch.stft`. This may not affect the final synthesis, and I will give it a try later.

then? have you tried?

**Comment by ease-zh** [comment]
@a897456 Not yet. I've been busy with something else.
